### ISSUE TITLE
Fast VL-PLM

### ISSUE THREAD

**Issue by PeizeSun** [xiaofeng94/VL-PLM#4]
Hi, thanks for your providing this simple and efficient work !

I'm quite interested to your attempt to improve time efficiency of generating pseudo labels. Would you mind sharing the code of Fast VL-PLM ?

**Comment by xiaofeng94** [comment]
Hey, thanks for your interests in our work, but we do not plan to release all the code for Fast VL-PLM. You might implement it by referring to the following code. This sample function can accelerate the generation of [image_features](https://github.com/xiaofeng94/VL-PLM/blob/main/tools/utils.py#L240)

```
def encode_image_fullBox(image_encoder, images, box_list, spatial_scale=1/32):
    device = next(image_encoder.parameters()).device

    if isinstance(image_encoder, ModifiedResNet):
        x = CLIP_RNx_main_forward(image_encoder, images)  # NCHW

        # one by one process
        orginPE = image_encoder.attnpool.positional_embedding
        orginPE_7x7 = orginPE[1:].reshape(7,7,-1)

        # only for batch size == 1
        curFeatMap = x  # N(1),C,H,W
        iH, iW = curFeatMap.shape[2:]
        iH, iW = np.floor([iH/spatial_scale, iW/spatial_scale]).astype(int)
        assert(curFeatMap.shape[0]==1), 'not supported for multiple images'

        globClsEmb = curFeatMap.flatten(start_dim=-2, end_dim=-1).mean(dim=-1, keepdim=True).permute(2, 0, 1)  # 1xN(1)xC

        boxFeatList = list()
        boxList = box_list
        for bx_idx, curBox in enumerate(boxList):
            curBox = scale_box(curBox, 1, max_H=iH, max_W=iW)
            box_x0, box_y0, box_x1, box_y1 = np.array(curBox) * spatial_scale

            box_top = min(max(int(np.floor(box_y0)), 0), curFeatMap.shape[2]-1) # y or vertical
            box_left = min(max(int(np.floor(box_x0)), 0), curFeatMap.shape[3]-1) # x or horizontal
            box_height = int(np.ceil(box_y1) - box_top)
            box_width = int(np.ceil(box_x1) - box_left)

            # if box_height > 7 and box_width > 7:
            if False:
                # print('box_height, box_width:', box_height, box_width)
                cropFeat = curFeatMap[:, :, box_top:box_top+box_height+1, box_left:box_left+box_width+1] # 1,C,box_height,box_width
                _, _, fH, fW = cropFeat.shape
                try:
                    newPE = F.interpolate(orginPE_7x7.permute(2,0,1).unsqueeze(0), [fH, fW], mode='bilinear')[0].permute(1,2,0) # fH x fW x dim
                    newPE = torch.cat([orginPE[0:1, :], newPE.flatten(start_dim=0, end_dim=1)], dim=0)  # (1 + fH x fW) x dim
                except:
                    import ipdb
                    ipdb.set_trace()
                    print("An exception occurred")

                boxFeat = my_attn_pool2d_forward(image_encoder.attnpool, cropFeat, pe=newPE, cls_emb=globClsEmb)
                # boxFeat = my_attn_pool2d_forward(image_encoder.attnpool, cropFeat, pe=None, cls_emb=globClsEmb)
            else:
                boxTR = torch.as_tensor(np.array(curBox), dtype=curFeatMap.dtype).reshape(1, -1).to(curFeatMap.device)
                boxFeat = torchvision.ops.roi_align(curFeatMap, [boxTR], output_size=7, spatial_scale=spatial_scale)
                boxFeat = image_encoder.attnpool(boxFeat)

            boxFeatList.append(boxFeat)

        out = torch.cat(boxFeatList, dim=0)

    else:
        print('non support image encoder')
        return

    return out
```

**Comment by PeizeSun** [comment]
Thanks a lot :)
